transmat/zip/zip_unpack.go

```diff
@@ -0,0 +1,202 @@
+					{"ziphdr", fmt.Sprintf("%#v", f)},
+				},
+			})
+			continue
+		}
+		if haltMe != nil {
+			return api.WareID{}, api.WareID{}, haltMe
+		}
+		if strings.HasPrefix(fmeta.Name.String(), "..") {
+			return api.WareID{}, api.WareID{}, Errorf(rio.ErrWareCorrupt, "corrupt zip: paths that use '../' to leave the base dir are invalid")
+		}
+
+		// Infer parents, if necessary.  The zip format should not allow implicit dirs, but we allow
+		// it for tars, so why not here.
+		for _, parent := range fmeta.Name.SplitParent() {
+			// If we already initialized this parent, superb; move along.
+			if _, exists := dirs[parent]; exists {
+				continue
+			}
+			// If we're missing a dir, conjure a node with defaulted values.
+			log.DirectoryInferred(mon, parent, fmeta.Name)
+			conjuredFmeta := fshash.DefaultDirMetadata()
+			conjuredFmeta.Name = parent
+			prefilterBucket.AddRecord(conjuredFmeta, nil)
+			filters.ApplyUnpackFilter(filt, &conjuredFmeta)
+			filteredBucket.AddRecord(conjuredFmeta, nil)
+			dirs[conjuredFmeta.Name] = struct{}{}
+			if err := fsOp.PlaceFile(afs, conjuredFmeta, nil, false); err != nil {
+				return api.WareID{}, api.WareID{}, Errorf(rio.ErrInoperablePath, "error while unpacking: %s", err)
+			}
+		}
+
+		// Apply filters.
+		//  ... uck, to one copy of the meta.  We can't add either to their buckets
+		//  until after the file is placed because we need the content hash.
+		filteredFmeta := fmeta
+		//  The filter may reject things by returning an error;
+		//   or, instruct us to ignore things by setting the type to invalid.
+		if err := filters.ApplyUnpackFilter(filt, &filteredFmeta); err != nil {
+			return api.WareID{}, api.WareID{}, err
+		}
+		if fmeta.Type == fs.Type_Invalid {
+			// skip placing that file and continue processing...
+			//  but *do* still record it in the prefilter bucket for hashing.
+			//  (n.b. currently only non-files are ever ejected like this.)
+			prefilterBucket.AddRecord(fmeta, nil)
+			continue
+		}
+
+		// Place the file.
+		switch fmeta.Type {
+		case fs.Type_File:
+			r, err := f.Open()
+			if err != nil {
+				return api.WareID{}, api.WareID{}, Errorf(rio.ErrInoperablePath, "error while unpacking: %s", err)
+			}
+			reader := &util.HashingReader{R: r, Hasher: sha512.New384()}
+			if err = fsOp.PlaceFile(afs, filteredFmeta, reader, false); err != nil {
+				return api.WareID{}, api.WareID{}, Errorf(rio.ErrInoperablePath, "error while unpacking: %s", err)
+			}
+			prefilterBucket.AddRecord(fmeta, reader.Hasher.Sum(nil))
+			filteredBucket.AddRecord(filteredFmeta, reader.Hasher.Sum(nil))
+			continue
+		case fs.Type_Symlink:
+			buf := new(bytes.Buffer)
+			r, err := f.Open()
+			if err != nil {
+				return api.WareID{}, api.WareID{}, Errorf(rio.ErrInoperablePath, "error while unpacking: %s", err)
+			}
+			buf.ReadFrom(r)
+			fmeta.Linkname = buf.String()
+			filteredFmeta.Linkname = fmeta.Linkname
+		case fs.Type_Dir:
+			dirs[fmeta.Name] = struct{}{}
+		default:
```

Comment:
Oh, boy, this took me a couple of reads.  I think it does the right thing but it's getting pretty alarmingly clever with the switch syntax.

Either a couple comments about intentionally falling out to the post-switch statements (or possibly better, putting those PlaceFile and two bucket operations in a helper function? or some other abstraction like that) would make this a lot more readable.